sdk/nodejs/appservice/zMixins.ts

```diff
@@ -415,6 +515,128 @@ export class CallbackFunctionApp<C extends Context<R>, E, R extends Result> exte
         this.plan = plan;
         this.zipBlob = zipBlob;
+
+        const routePrefix = args.hostSettings 
+            && args.hostSettings.extensions 
+            && args.hostSettings.extensions.http 
+            && args.hostSettings.extensions.http.routePrefix;
+        const rootPath = routePrefix === "" ? "" : `${routePrefix === undefined ? "api" : routePrefix}/`;
+
+        this.endpoint = pulumi.interpolate`https://${this.defaultHostname}/${rootPath}`;        
+    }
+}
+
+/**
+  * A CallbackFunctionApp is a special type of azure.appservice.FunctionApp that can be created out
+  * of an actual JavaScript function instance.  The function instance will be analyzed and packaged
+  * up (including dependencies) into a form that can be used by Azure Functions. See
+  * https://github.com/pulumi/docs/blob/master/reference/serializing-functions.md for additional
+  * details on this process.
+ */
+export class CallbackFunctionApp<C extends Context<R>, E, R extends Result> extends CombinedFunctionApp {
+    constructor(name: string, bindings: pulumi.Input<BindingDefinition[]>,
+                args: CallbackFunctionAppArgs<C, E, R>, opts: pulumi.CustomResourceOptions = {}) {
+        const functionsArgs = {
+            ...args,
+            functions: [<Function>{
+                name,
+                bindings,
+                callback: args,
+            }]
+        };
+        const archive = pulumi.output(produceDeploymentPackage(functionsArgs)).apply(m => new pulumi.asset.AssetArchive(m));
+
+        super(name, { ...args, archive }, opts);
+    }
+}
+
+/**
+ * A base component for custom Function App components. It is required to provide a common parent for all resources,
+ * so that they are logically groupped under the same root in the Pulumi resource tree.
+ */
+class ArchiveFunctionAppBase extends pulumi.ComponentResource {
+    /**
+     * Storage account where the FunctionApp's zipbBlob is uploaded to.
+     */
+    public readonly account: storageForTypesOnly.Account;
+    /**
+     * Storage container where the FunctionApp's zipbBlob is uploaded to.
+     */
+    public readonly container: storageForTypesOnly.Container;
+    /**
+     * The blob containing all the code for this FunctionApp.
+     */
+    public readonly zipBlob: storageForTypesOnly.ZipBlob;
+    /**
+     * The plan this Function App runs under.
+     */
+    public readonly plan: appservice.Plan;  
+    /**
+     * The plan this Function App runs under.
+     */
+    public readonly functionApp: appservice.FunctionApp;   
+    /** 
+     * Root HTTP endpoint of the Function App.
+     */
+    public readonly endpoint: pulumi.Output<string>;
+
+    constructor(type:string,
+                name: string,
+                args: ArchiveFunctionAppArgs,
+                opts: pulumi.ComponentResourceOptions = {}) {
+        super(type, name, undefined, opts);
+
+        const app = new CombinedFunctionApp(name, args, { parent: this });

```

Comment:
AFAICT, CombinedFunctionApp doesn't really serve much purpose.  It's just there to create all this underlying data, which you then export out of this type.  We technically could just use a normal non-resource object for CombinedFunctionApp.  Or just have a function that returned a pojo with this data.

Reply:
oh, i see.  you also have CallbackFuncitonApp extend this.  Ok. i need to write this all down :D

Reply:
I tried to outline this in https://github.com/pulumi/pulumi-azure/pull/264#issuecomment-501199117, let me know if you see better ways

Reply:
👍